### PR TITLE
[plugin] Cache command arguments to safely pass the command over JSON-RPC

### PR DESCRIPTION
**Pull request by vinokurig**
#### What it does

- Pass command to tree item in the tree-view-widget
- Cache command arguments to safely pass the command over JSON-RPC:
Currently tree item command doesn't provide arguments at all, they are overwritten to `id`:
https://github.com/theia-ide/theia/blob/f718cb3461eaf1a7de90c679d31cdbeafe485708/packages/plugin-ext/src/plugin/tree/tree-views.ts#L247-L249
The converter caches the command in the plugin side so the tree-view passes the command with id in arguments. This prevents `JSON circular structure` error and real arguments are available when the command is executed in the plugin side.
#### How to test

1. Clone sample plugin: https://github.com/vinokurig/Test.git and compile it
2. Start the plugin as a Hosted Plugin
3. Open the tree-view (list icon on the left panel):
screenshot-localhost-3030-2019 08 16-18-01-11

4. click the `label` item
**See** A message from argument's field is shown.

The argument object contains a reference to itself: https://github.com/vinokurig/Test/blob/cdd7548fd1d1d3f22ba263017a80ef8a6192ac52/src/nodeDependencies.ts#L18-L20
so the command was safely converted and passed over JSON-RPC without an error.
#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 1]
```diff
@@ -123,3 +130,51 @@ export class CommandRegistryImpl implements CommandRegistryExt {
         this.argumentProcessors.push(processor);
     }
 }
+
+export class CommandsConverter {
```
Comment (akosyakov) [inline packages/plugin-ext/src/plugin/command-registry.ts]: make sure to file a CQ for all copied code, it's copied and modified from here: https://github.com/TypeFox/vscode/blob/70b8db24a37fafc77247de7f7cb5bb0195120ed0/src/vs/workbench/api/common/extHostCommands.ts#L221-L263

please add a comment in the code and MS copyright under ours

cc @theia-ide/ip-help

Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: Added a copyright and a comment, opened a CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=20680

**Comment by vinokurig** [comment]
fixes https://github.com/eclipse/che/issues/13956

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 2]
```diff
@@ -123,3 +130,57 @@ export class CommandRegistryImpl implements CommandRegistryExt {
...
+    toSafeCommand(command: theia.Command, disposables: DisposableCollection): theia.Command {
+        if (!this.isCommandRegistered) {
+            this.commands.registerCommand({ id: this.safeCommandId }, this.executeSafeCommand, this);
+            this.isCommandRegistered = true;
+        }
+
+        const result: theia.Command = {};
+        Object.assign(result, command);
+
+        if (command.command && command.arguments && command.arguments.length > 0) {
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/command-registry.ts]: If I understand this correctly, this will make it impossible to invoke commands registered in one extension host from a different extension host (because the arguments are not passed through the front end). While this is a non-issue in Theia proper, it is a breaking change for Che plugins. Maybe we should run this change by the chedev mailing list first.

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 3]
```diff
@@ -123,3 +130,57 @@ export class CommandRegistryImpl implements CommandRegistryExt {
...
+        this.safeCommandId = `theia_safe_cmd_${Date.now().toString()}`;
+        this.commands = commands;
+        this.isCommandRegistered = false;
+    }
+
+    /**
+     * Convert to a command that can be safely passed over JSON-RPC.
+     */
+    toSafeCommand(command: theia.Command, disposables: DisposableCollection): theia.Command {
+        if (!this.isCommandRegistered) {
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/command-registry.ts]: "isSafeCommandRegistered" would make it clear it's not the command that is being converted that needs to be registered.

**Comment by tsmaeder** [comment]
@vinokurig is there an issue this PR fixes? I can't find a reference in the PR description.

**Comment by vinokurig** [comment]
@tsmaeder 
> @vinokurig is there an issue this PR fixes? I can't find a reference in the PR description.

https://github.com/theia-ide/theia/pull/5961#issuecomment-522453481

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 2 above)
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: It won't affect commands without arguments. This logic just re-register commands that have arguments. I don't see any reasons why a command won't be invoked.

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 3 above)
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: done

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 2 above)
Comment (tsmaeder) [reply packages/plugin-ext/src/plugin/command-registry.ts]: Ok, I see that "executeSafeCommand" delegates to the command registry for execution. So plugins have the option to have "unsafe" arguments (and execute locally) or safe arguments that will go through the "executeCommand" API call back to the front end.

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 4]
```diff
@@ -34,10 +35,16 @@ export class CommandRegistryImpl implements CommandRegistryExt {
     private readonly commands = new Set<string>();
     private readonly handlers = new Map<string, Handler>();
     private readonly argumentProcessors: ArgumentProcessor[];
+    private readonly commandsConverter: CommandsConverter;
```
Comment (tsmaeder) [inline packages/plugin-ext/src/plugin/command-registry.ts]: CommandRegistryImpl never does anything with the "commandsConverter". IMO, the converter should be owned by 
"TreeViewExtImpl", not the command registry.

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 2 above)
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: Actually commands will be executed with real arguments that was saved in the plugin side:
https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/command-registry.ts#L179
if the command was registered by the plugin it will be executed locally:
https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/command-registry.ts#L98-L99
Safe command is needed when the tree-view passes items to the front-end to prevent `JSON circular structure`error: https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/tree/tree-views.ts#L263

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 4 above)
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: This can be used by other APIs.

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 1 above)
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: @akosyakov The CQ has been approved!

**Comment by akosyakov** [comment]
@olexii4 Does your approve means that changes looks good and you tested that it behaves well? Please provide such comments to make it clear for other committers that PR is well reviewed already. We included it in review guidelines for such reasons: https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#justifiying-approve

File: packages/plugin-ext/src/plugin/command-registry.ts (hunk 1 above)
Comment (akosyakov) [reply packages/plugin-ext/src/plugin/command-registry.ts]: cool, thank you!

**Comment by akosyakov** [comment]
@vinokurig If @tsmaeder is fine with changes and @olexii4 approve means https://github.com/theia-ide/theia/pull/5961#issuecomment-522958483 you can proceed

**Comment by akosyakov** [comment]
I cannot verify, it fails for me:
```
root ERROR Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/workspace/theia/node_modules/express/lib/response.js:767:10)
    at ServerResponse.send (/workspace/theia/node_modules/express/lib/response.js:170:12)
    at /workspace/theia/packages/plugin-ext/lib/hosted/node/plugin-reader.js:112:37
    at /workspace/theia/node_modules/express/lib/response.js:430:22
    at SendStream.onend (/workspace/theia/node_modules/express/lib/response.js:1047:5)
    at SendStream.emit (events.js:189:13)
    at ReadStream.onend (/workspace/theia/node_modules/send/index.js:822:10)
    at ReadStream.emit (events.js:194:15)
    at endReadableNT (_stream_readable.js:1125:12)
root INFO [hosted-plugin: 1934] PLUGIN_HOST(1934): PluginManagerExtImpl/init()
root INFO [hosted-plugin: 1934] PLUGIN_HOST(1934): initializing(custom-view-samples@0.0.1 with /workspace/theia/packages/plugin-ext-vscode/lib/node/plugin-vscode-init.js)
root INFO [hosted-plugin: 1934] PLUGIN_HOST(1934): PluginManagerExtImpl/loadPlugin(/workspace/theia/plugins/Test/out/extension.js)
root ERROR Uncaught Exception:  Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
root ERROR Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/workspace/theia/node_modules/express/lib/response.js:767:10)
    at ServerResponse.send (/workspace/theia/node_modules/express/lib/response.js:170:12)
    at /workspace/theia/packages/plugin-ext/lib/hosted/node/plugin-reader.js:112:37
    at /workspace/theia/node_modules/express/lib/response.js:430:22
    at SendStream.onend (/workspace/theia/node_modules/express/lib/response.js:1047:5)
    at SendStream.emit (events.js:189:13)
    at ReadStream.onend (/workspace/theia/node_modules/send/index.js:822:10)
    at ReadStream.emit (events.js:194:15)
    at endReadableNT (_stream_readable.js:1125:12)
root ERROR Uncaught Exception:  Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
root ERROR Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:470:11)
    at ServerResponse.header (/workspace/theia/node_modules/express/lib/response.js:767:10)
    at ServerResponse.send (/workspace/theia/node_modules/express/lib/response.js:170:12)
    at /workspace/theia/packages/plugin-ext/lib/hosted/node/plugin-reader.js:112:37
    at /workspace/theia/node_modules/express/lib/response.js:430:22
    at SendStream.onend (/workspace/theia/node_modules/express/lib/response.js:1047:5)
    at SendStream.emit (events.js:189:13)
    at ReadStream.onend (/workspace/theia/node_modules/send/index.js:822:10)
    at ReadStream.emit (events.js:194:15)
    at endReadableNT (_stream_readable.js:1125:12)
```

I don't think that it is caused by this PR. It seems to be something broken with plugin systems generally now. The same issue were observed in my Monaco PR and in https://github.com/theia-ide/theia/issues/5967

**Comment by akosyakov** [comment]
I will do a PR promptly, found an issue.

**Comment by akosyakov** [comment]
@vinokurig please review https://github.com/theia-ide/theia/pull/6002, after that we can rebase your PR and test it

**Comment by vinokurig** [comment]
@akosyakov Re-based the branch.
While testing I've discovered that If the tree has more than one node with nested children, the command will work only for children of currently expanded node. That's because the internal commands map refreshes on each children request:
https://github.com/theia-ide/theia/blob/191bb6d02a841fe1f448809bfa1c64f032d00edb/packages/plugin-ext/src/plugin/tree/tree-views.ts#L177
In the vscode command is registered only when the tree node is fetched so they can refresh the commands map on the refresh tree request. In Theia tree nodes are always requested by the tree, so the commands map is refreshed on the children request to not to overload the commands map. I think we need to change the behavior of the tree like in the vscode way: https://github.com/theia-ide/theia/issues/6007

**Comment by akosyakov** [comment]
I see, please put it the description of #6007 as well.

**Comment by akosyakov** [comment]
> click the label item
See A message from argument's field is shown

@vinokurig I'm trying but clicking on the label item does not do anything. Where should it appear?

**Comment by akosyakov** [comment]
It only works if i click `Refresh` in the view or reload the browser. If i do `Reset Workbench Layout` it does not work anymore.

**Comment by vinokurig** [comment]
@akosyakov 
> It only works if i click Refresh in the view or reload the browser. It does not work for me on first start for some reasons. You can also reproduce it by removing the focus from the view (i.e. open the navigator) and then do Reset Workbench Layout command.

Fixed, it was a bug in the `tree-view-widget.tsx`

**Comment by vinokurig** [comment]
@akosyakov 
You can check it with https://github.com/microsoft/vscode-extension-samples/tree/master/tree-view-sample as well.
The commands from the Node Dependencies tree works: the web page of the dependency is opened by clicking the tree node.
**NOTE** The commands works only from the latest opened parent, If you click on node from previously opened parent it won't work. Can we go with it and than fix it in the https://github.com/theia-ide/theia/issues/6007 ?

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx [hunk 5]
```diff
@@ -141,13 +141,14 @@ export class PluginTree extends TreeImpl {
             }, update);
         }
         if (TreeViewNode.is(node)) {
-            return Object.assign(node, update);
+            return Object.assign(node, update, { command: item.command });
```
Comment (akosyakov) [inline packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx]: @vinokurig command cannot be applied to composite elements? just checking

**Comment by akosyakov** [comment]
Tested it with VS Code as well, let me try again.

File: packages/plugin-ext/src/plugin/command-registry.ts [hunk 6]
```diff
@@ -123,3 +130,57 @@ export class CommandRegistryImpl implements CommandRegistryExt {
         this.argumentProcessors.push(processor);
     }
 }
+
+/*---------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply packages/plugin-ext/src/plugin/command-registry.ts]: done

File: packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx (hunk 5 above)
Comment (vinokurig) [reply packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx]: yes, according to previous logic that was lost:
https://github.com/theia-ide/theia/blob/93205109fa4017286ec0a441b0250b74b36c6495/packages/plugin-ext/src/main/browser/view/tree-views-main.tsx#L206-L218
command is set to `fileNode`, bit is not set to `folderNode`:
https://github.com/theia-ide/theia/blob/93205109fa4017286ec0a441b0250b74b36c6495/packages/plugin-ext/src/main/browser/view/tree-views-main.tsx#L189-L204

Comment (akosyakov) [reply packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx]: I meant whether it should be set on the composite according to VS Code expectations? If so it would be nice to adjust it as well. It could be done separately, but needs an issue at least.

Comment (vinokurig) [reply packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx]: Raised an issue with question https://github.com/theia-ide/theia/issues/6025

**Comment by akosyakov** [comment]
@vinokurig please merge, let's tackle remaining 2 issues separately

**Comment by olexii4** [comment]
I tested these changes inside Eclipse CHE

I built the new image "olexii4dockerid/che-theia:next" 
https://github.com/eclipse/che-theia
```
./build.sh --build-args:GITHUB_TOKEN=${GITHUB_TOKEN},--tag:next --branch:che-13956 --git-ref:refs\\/heads\\/master
```
and push it to upstream.
Create a new workspace from the devfile 
https://gist.githubusercontent.com/olexii4/99aa676393255e70e03f70b0cafe5bde/raw/6e838186630eb64c272d9bf698666569610d815a/devfile.yaml
Open 'my Workspace' and try open a link from the tree. 

Screenshot from 2019-08-26 19-31-35
